src/AutoRest.CSharp/AutoRest/Plugins/CSharpGen.cs

```diff
@@ -75,10 +75,19 @@ public async Task<GeneratedCodeWorkspace> ExecuteAsync(Task<CodeModel> codeModel
 
             foreach (var client in context.Library.Clients)
             {
-                var codeWriter = new CodeWriter();
-                clientWriter.WriteClient(codeWriter, client, context.Configuration);
+                var credentialTypes = context.Configuration.CredentialTypes;
+                if (credentialTypes.Contains("TokenCredential", StringComparer.OrdinalIgnoreCase) ||
+                    credentialTypes.Contains("AzureKeyCredential", StringComparer.OrdinalIgnoreCase))
+                {
+                    var codeWriter = new CodeWriter();
+                    ClientWriter.WriteClientOptions(codeWriter, client, context);
+                    project.AddGeneratedFile($"{client.Type.Name}Options.cs", codeWriter.ToString());
```

Comment:
Would this result in a clientoptions type per client? We only need 1 for the entire library.

Reply:
Updated to generate on client options for the entire library.